### PR TITLE
Expand Floquet calibration to arbitrary FSim gates

### PR DESCRIPTION
**Pull request by fedimser**
* Added a method try_convert_gate_to_fsim which can convert to PhaseCalibratedFSimGate wide range of gates, including sqrt(iSWAP), Sycamore and Controlled-Z gates.
* Used this method (instead of try_convert_sqrt_iswap_to_fsim) as default gate translator in all methods in workflow.py.

As a result, compensation step Floquet calibration now works not only for sqrt(iSWAP), but also for:
* all FSim gates (including Sycamore gate);
* PhasedFSim gates (such that zeta=gamma=0); 
* all ISwapPow gates;
* all PhasedISwapPow gates;
* CZPow gates without global phase (including CZ).

This change doesn't affect characterization step the Floquet calibration.

### CODE REVIEW

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/phased_fsim.py:1043]: If you look at my comment below, it would be good to split `prepare_` and `make_` methods to use different converters. However, because of that, it would also be good that whatever is created with `prepare_` stays consistent with `make_` so that you could always reuse the results.

Maybe you could always call the old method first, and do the additional gate resolution only after the result is `None`?

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/phased_fsim.py:1054]: If you look at the previous method, the code is trying to unify the sqrt_iswap ad sqrt_iswap inverse into a shared `engine_gate`. This could also be done here, the `PhasedFSimGate` is not an unique representation and there is a freedom between `theta` and `chi`.

It is important to do so because of the way characterizations work: they are applied to the entire moment and if we could unify two different moments into a single characterization then we'll save a lot of time. Characterizations take considerable amount of time to do.

(The same applies to IswapPowGate, CZPowGate and PhasedIswapPowGate cases).

Please reach out to me over email if you want to have some references about the math behind this. I have some notes on this in multiple places and I can summarize it for you.

File: cirq-google/cirq_google/calibration/phased_fsim_test.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/phased_fsim_test.py:880]: Could you expand the `test_phase_calibrated_fsim_gate_compensated` and `test_try_convert_sqrt_iswap_to_fsim_converts_correctly` with the newly supported gates, so that we have a checks that test the resulting untaries as well?

(Tests based on `assert_allclose_up_to_global_phase` or `np.allclose(cirq.unitary(fsim), expected_unitary)`)

File: cirq-google/cirq_google/calibration/phased_fsim_test.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/phased_fsim_test.py:912]: How does this work? I don't see the SYC gate create for the new converter?

**Review by mrwojtek — COMMENTED** [review]
Thank you Dmytro for doing this! The concepts contained here are a little bit complex. I've tried to explain some of them but feel free to ping me if some of my comments are not clear for you.

File: cirq-google/cirq_google/calibration/phased_fsim_test.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/phased_fsim_test.py:912]: This is because SycamoreGate inherits from FSimGate, so it is handled by the branch "if isinstance(gate, cirq.FSimGate)".
I added an explicit branch so it would produce cirq.FSimGate, not SycamoreGate - but I don't think it's necessary.

File: cirq-google/cirq_google/calibration/workflow_test.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/workflow_test.py:1388]: I changed the test to use gates of the same kind in the same moment.

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/phased_fsim.py:1054]: I think I understand what needs to be done here. According to my calculations, matrix of `PhaseCalibratedFSimGate(FSimGate(theta, phi), p)` is:

```
1, 0, 0, 0
0, cos(theta), exp(2*pi*p*i)*(-sin(theta)), 0
0, exp(-2*pi*p*i)*(-sin(theta)), cos(theta), 0
0, 0, 0, exp(i*phi)
```

So the transformation `theta -> -theta; p -> p+ 0.5` would not change the matrix. Therefore, by using equivalent tranformations of parameters we can always put theta in the range [0, pi).
I will do that and it will be not worse than what is done `try_convert_sqrt_iswap_to_fsim`, because sqrt(iSWAP) and its inwerse both will be mapped to the same engine gate.

File: cirq-google/cirq_google/calibration/workflow.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/workflow.py:75]: I reverted all changed except in two `make_` methods.

File: cirq-google/cirq_google/calibration/workflow_test.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/workflow_test.py:1401]: Added this check.
It forced me to rewrite `test_zeta_chi_gamma_calibration_for_moments`, because now each gate has its own engine gate, and only that engine gate can be passed in `characterizations`.

File: cirq-google/cirq_google/calibration/workflow_test.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/workflow_test.py:1423]: That would require expanding engine_simulator to support gates other than sqrt_iswap...
I removed this entire test for now, because it's wrong anyway. Let's chat and decide how to write this test.

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/phased_fsim.py:1044]: Actually, after correction I did today (which ensures theta is in range [0, np.pi)), calling `try_convert_gate_to_fsim` is equivalent to calling `try_convert_sqrt_iswap_to_fsim` and then calling `try_convert_gate_to_fsim` if first call returned `None`.

**Comment by fedimser** [comment]
Hi Wojtek, I resolved most of your comments, and rewrote `try_convert_gate_to_fsim` so it tries to make engine gate as "canonical" as possible, in paricular now theta is in range [0, pi).
I am still not sure how to write good integration test for the new compensation - let's discuss that. Should i expand engine simulator to support arbitrary imperfect FSim gates (and not only sqrt(iSWAP)).

**Comment by fedimser** [comment]
I added a test utilizing the engine simulator. Please take another look.

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/engine_simulator.py:309]: The other method is called `create_from_dictionary_sqrt_iswap`, maybe `create_from_dictionary` would be more consistent with that?

Could you also move this method up, closer to the original one?

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/engine_simulator.py:311]: It would be good to keep the argument names consistent, with this one just being `parameters`. There are also other arguments in the original method like `simulator` and `ignore_`, can you add them here as well? This is externally visible method.

In a follow up we can remove all the logic from `create_from_dictionary_sqrt_iswap` and just delegate to this method but in order to do so, this needs to be a generalization of the other one.

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/engine_simulator.py:336]: Don't abbreviate in externally visible error messages (params -> parameters)

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/engine_simulator.py:342]: The original method had a check here which tested if all the pairs are sorted. This prevents providing the incorrect input where swapped gate is very different from the non-swapped gate. I think either one of those checks should be here, to prevent incorrect simulator state.

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/engine_simulator.py:337]: This is a `ValueError`, not an assertion. Assertions are used for checking internal workflows and what you're doing here is checking the user input.

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/engine_simulator.py:333]: Could you keep the logic that I had in the previous method (`pair_parameters = PhasedFSimCharacterization(**pair_parameters)`). This is very convenient in tests.

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/phased_fsim.py:1076]: Can you change `p` to `phase_exponent`?

File: cirq-google/cirq_google/calibration/phased_fsim_test.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/phased_fsim_test.py:881]: We typically don't underscore inner functions

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/phased_fsim.py:1000]: Could you add a docstring to mention that this is cirq unitary protocol implementation?

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (mrwojtek) [reply cirq-google/cirq_google/calibration/phased_fsim.py:1050]: This is a strong requirement for us because if it fails, the characterization and compensation phases are different. Could you add e test specifically to test this condition?

File: cirq-google/cirq_google/calibration/workflow_test.py
Comment (mrwojtek) [reply cirq-google/cirq_google/calibration/workflow_test.py]: I think this is done!

**Review by mrwojtek — COMMENTED** [review]
This is getting much simpler, a good sign.

File: cirq-google/cirq_google/calibration/engine_simulator_test.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/engine_simulator_test.py:444]: I think this needs a test similar to the `test_from_dictionary_sqrt_iswap_simulates_correctly` test.

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/engine_simulator.py]: Renamed gates_dict->parameters and added simulator.
What do you mean by `ignore_`. Are you referring to `ideal_when_missing_gate` and `ideal_when_missing_parameter`? Do we really need them here? 

As far as i'm concerned, I'm adding this factory method just so I can write a test, so I am passing only these parameters which I need. In my test I want to require that all FSim-like gates in circuit must be explicitly specified in engine simulator.

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/engine_simulator.py]: Fixed.

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/engine_simulator.py]: Done.

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/engine_simulator.py]: Added the check.

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/engine_simulator.py]: I removed that cjeck and instead added PhasedFSimCharacterization(**pair_parameters), as you suggested.

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/engine_simulator.py]: Done.

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/phased_fsim.py]: Done.

File: cirq-google/cirq_google/calibration/phased_fsim_test.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/phased_fsim_test.py:881]: Done.

File: cirq-google/cirq_google/calibration/engine_simulator_test.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/engine_simulator_test.py:490]: Added a test.

**Comment by fedimser** [comment]
Hi @mrwojtek, I addressed your comments and added more tests. PTAL.

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (mrwojtek) [reply cirq-google/cirq_google/calibration/phased_fsim.py:1051]: After our discussion yesterday this is no longer relevant, since it will be a single function soon.

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/phased_fsim.py:1082]: I just realized this property is shared among all gates. Should it be a check shared for all the gates. Does it actually matter at all, why CZ is special?

File: cirq-google/cirq_google/calibration/workflow_test.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/workflow_test.py:1405]: Could you use cirq.SycamoreGate here?

File: cirq-google/cirq_google/calibration/workflow_test.py
Comment (mrwojtek) [inline cirq-google/cirq_google/calibration/workflow_test.py:1404]: Could you use cirq.CZ gate here and below?

**Review by mrwojtek — COMMENTED** [review]
Three minor issues and one question which maybe you could clarify for me and we are done I think.

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/phased_fsim.py:1082]: global_shift is not shared among all gates, it's property of the EigenGate: https://quantumai.google/reference/python/cirq/ops/EigenGate
In this method EigenGates are: ISwapPowGate and CZPowGate.

I just did the math and turned out that specifying non-zero global_shift (for any EigenGate) is equivalent to multiplying  gate's matrix by a scalar exp(pi*i*global_shift*exponent).

It does matter, because in case when exp(pi*i*global_shift*exponent) != 1, the input gate is not PhasedFSimGate.

So this is what I will do:
1. Change check to np.isclose(np.exp(np.pi*1j*gate.global_shift*gate.exponent), 1.0);
2. Add a test for non-trivial case (e.g. global_shift=1/3, exponent=3/2); I doubt this will happen on practice, but we will be able to clain that this method translates all gates which can be translated;
3. Do the same check for ISwapPowGate, but not for other gates.

Comment (mrwojtek) [reply cirq-google/cirq_google/calibration/phased_fsim.py:1082]: Thanks for diving into that!

File: cirq-google/cirq_google/calibration/engine_simulator.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/engine_simulator.py:253]: Done.

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/phased_fsim.py]: This is now fixed.

File: cirq-google/cirq_google/calibration/workflow_test.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/workflow_test.py:1404]: I can't, because cirq.CZ != FSimGate(theta=0, phi=np.pi), but gate translator converts all gates to FSimGate. So  all gates in this map must be FSimGate, otherwise there will be no match.

File: cirq-google/cirq_google/calibration/workflow_test.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/workflow_test.py:1405]: Yes, I can - but only because SycamoreGate inherits from FSimGate.
Used cirq_google.SYC.

**Comment by fedimser** [comment]
I addressed issue with global_phase and other suggestions. 
@mrwojtek

File: cirq-google/cirq_google/calibration/workflow_test.py
Comment (mrwojtek) [reply cirq-google/cirq_google/calibration/workflow_test.py:1404]: Yes, you're right. If this code will be used heavily in the future we might think about normalizing all gates to a common form inside the "create_from_dictionary" by using the converter itself. (Let's postpone it for now.)

**Review by mrwojtek — APPROVED** [review]

**Comment by mrwojtek** [comment]
@balopat can we have approval from code owners from this?

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (tanujkhattar) [inline cirq-google/cirq_google/calibration/phased_fsim.py:1095]: This assumes that the gate to convert is not parameterized, is this assumption correct? If yes, please document it and raise error if not valid. If not, then please update the modular arithmetic to use https://github.com/quantumlib/Cirq/blob/0311e1746acdb4519f334c421a90e15ff3035891/cirq-google/cirq_google/common_serializers.py#L51

**Review by tanujkhattar — APPROVED** [review]
LGTM % comment

File: cirq-google/cirq_google/calibration/phased_fsim.py
Comment (fedimser) [reply cirq-google/cirq_google/calibration/phased_fsim.py:1095]: Hi. I didn't realize that some gates have cirq.value.TParamVal as theoir parameters. I assume you are referring to that.

Since output of this function (PhaseCalibratedFSimGate and FSimGate) can have only `float` parameters, we can't support parameterized gates.

I'm going to add a check that all parameters are `float`s, and otherwise return `None`.

Comment (tanujkhattar) [reply cirq-google/cirq_google/calibration/phased_fsim.py:1105]: The implementation of `FSim` gate does allow the parameters to be symbols (though the types in constructor are not correctly updated).  Eg:

https://github.com/quantumlib/Cirq/blob/12530f1531e5da663a6b172b1967763e08819912/cirq-core/cirq/ops/fsim_gate.py#L39

Hence, the following code is perfectly valid in Cirq. 

```python
fsim = cirq.FSimGate(theta = sympy.Symbol('theta'), phi = sympy.Symbol('phi'))
cirq.Circuit(fsim(*cirq.LineQubit.range(2)))
```

I guess what you mean is that for the specific context here, you do not wish to support parameterized gates? You can use `cirq.is_parameterized()` protocol to check whether the input gate is parameterized or not.

Comment (fedimser) [reply cirq-google/cirq_google/calibration/phased_fsim.py:1098]: Thanks, I used `cirq.is_paramterized()`.

**Pull request merged**
